### PR TITLE
Remove `_base_vt`

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #193679
* #193678
* #193677
* #193676
* #193675
* #193674
* #193673

Consider the following example:

```python
class MyList(list):
    def append(self, item):
        super().append(item)
```

Dynamo represented this inheritance by storing a `_base_vt` object on `UserDefinedObjectVariable` (UDOV), pointing to a `ListVariable`. This composition-based approach had limitations:

* Storage lived in a separate location (`_base_vt`), while dispatch logic remained on the UDOV itself.
* The `_base_vt` object was unaware it was serving as the base for a user-defined object, which broke identity/equality semantics in edge cases like:

```python
>>> class F(frozenset): ...
>>> f = F()
>>> id(f) == id(f.copy())
False
>>> h = frozenset()
>>> id(h) == id(h.copy())
True
```

Because the underlying `_base_vt` had no notion of being wrapped by a user-defined object, Dynamo's `.copy()` incorrectly returned the same identity in both cases.

Rather than composition, this PR uses inheritance to directly extend the behavior of the base type. The VT inheritance for a class `class MyList(list)` is `UserDefinedList` -> `UserDefinedObjectVariable` -> `ListVariable` -> `VariableTracker`. The reason for this choice is:

* `UserDefinedObjectVariable` is the `PyHeapType` and knows how to redirect any custom method calls to the appropriate user-defined implementations.
* `ListVariable` contains the builtin behavior for a list type.
* Inheritance gives a single MRO instead of two separate objects, so `isinstance` checks and attribute/method lookup fall through naturally to the builtin behavior without needing a separate `_base_vt` redirect.

Each commit in this stack migrates a specific user-defined object from composition to inheritance, but only the top commit ([#193679](https://github.com/pytorch/pytorch/pull/193679)) is relevant for CI. Before merging, I'll squash everything into a single commit.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Failures are related to https://github.com/pytorch/pytorch/issues/191579

**Comment by guilhermeleobas** [comment]
@claude please review this stack. Am I missing anything?

**Comment by anijain2305** [comment]
AI assisted code review

Reviewed the `_base_vt` composition -> multiple-inheritance refactor by reading the diff only (worktree isn't built, so nothing was executed). Findings, most severe first.

### 1. `_base_methods` for user exceptions is now almost empty — `str()`/`repr()` on `Exception` subclasses should graph-break
`torch/_dynamo/variables/user_defined.py:4430-4439`

`self._base_methods = exception_methods if isinstance(value, Exception) else base_exception_methods`

`exception_methods` is built from `Exception.__dict__` = `{__init__, __new__}` only. `__str__`, `__repr__`, `with_traceback`, `add_note`, `__reduce__`, `__setstate__` live in `BaseException.__dict__` and are distinct objects (`Exception.__init__ is not BaseException.__init__`). The old expression selected `base_exception_methods` for every exception — its `else` branch was dead — so this swaps one half of the set for the other rather than fixing it.

For `class MyErr(Exception): pass`, `repr(e)` -> `tp_repr_impl` (`user_defined.py:1965`) -> `inherits_base_slot("__repr__")` False -> `_maybe_lookup_method` (`user_defined.py:2250`) -> `BaseException.__repr__` is a wrapper descriptor not in `_base_methods` -> `unimplemented("C-implemented special method without VariableTracker model")`. Same for `str(e)` via `tp_str_impl:2000`, and the old `tp_repr_impl`/`tp_str_impl` overrides that routed these to `exc_vt` were deleted, so there is no fallback. The set likely wants to be the union of both.

### 2. `is_modified` is unconditionally true for dict/set subclasses with a non-local source
`torch/_dynamo/side_effects.py:740-747`, `dicts.py:340-344`, `sets.py:160-165`

`has_new_items()` returns `should_reconstruct_all or any(...)`, and `should_reconstruct_all` is `not is_from_local_source(source)` (`dicts.py:160`) — True for any global/closure-rooted dict or set subclass, mutated or not.

`install_dict_contains_guard` (`dicts.py:451`) and `install_set_contains_guard` (`sets.py:199`) early-return on `is_modified`, assuming a `DICT_KEYS_MATCH` guard was already installed by the mutation. With no mutation there is no such guard, so `k in G.my_dict_subclass` is baked into the compiled code unguarded. These objects also enter `_get_modified_vars()`, emitting the clear+repopulate replay on every compile. Previously this came from the composite `ValueMutationExisting.is_modified` flag, which was False when unmutated.

### 3. New exception `reconstruct` imports the class by module+name and re-runs `__init__`
`torch/_dynamo/variables/user_defined.py:4479-4508`

`load_import_from(self.exc_type.__module__, self.exc_type.__name__)` only works for exception classes that are module-level attributes; one defined inside a function fails at replay time in generated bytecode rather than graph-breaking. The call also re-executes the user's `__init__` (the previous UDOV path built via `base_cls.__new__`), and `__traceback__` is still not replayed.

### Minor
- `UserDefinedDictVariable._new_dict` (`user_defined.py:4692`) ignores its `items` argument and copies `self.items`; benign only because `dict_copy` is the sole caller today, unlike the sibling `_new_set`/`_new_list`.
- `_length_impl` (`user_defined.py:3002`) re-checks `inherits_base_slot("__len__")` that both callers already checked, and would call `super().sq_length_impl` on the `mp_` path.
- `call_base_method` (`user_defined.py:1901`) raises a bare `StopIteration` if no base class implements the slot.
- `UserDefinedConstantVariable.__init__` (`user_defined.py:4600`): `_constant_base` is never read, and the old "No constant base type found in MRO" assertion was dropped, so an unexpected type silently gets `_base_methods = None`.

### Test coverage
The test delta is thin for a refactor this broad: one repro test, one `expectedFailure` removal, a re-parametrization, and 11 deleted CPython expected-failure stubs. Nothing new covers subclasses that override base methods (`update`, `__setitem__`, `keys`) and must not be short-circuited by `tp_method` dispatch, `copy()`/slicing/set-ops returning plain builtins, the new `UserDefinedOrderedDictVariable`/`UserDefinedFrozensetVariable`, or guard/recompile behaviour for a globally-scoped dict subclass.

File: torch/_dynamo/side_effects.py [hunk 1]
```diff
@@ -740,10 +737,27 @@ def is_modified(self, item: VariableTracker) -> bool:
...
+        ):
+            return True
+
+        if (
+            isinstance(
+                item,
+                (variables.UserDefinedListVariable, variables.UserDefinedDequeVariable),
+            )
+            and item.source is not None
+            and item.source in self.mutated_sources
```
Comment (anijain2305) [inline torch/_dynamo/side_effects.py]: `self.mutated_sources` is not something we should depend upon yet. It is an afterthought to support nested compile region, and I am not sure if it is rock solid to depend upon here.

File: torch/_dynamo/side_effects.py [hunk 2]
```diff
@@ -1960,28 +1937,22 @@ def _codegen_attribute_mutation(ctx: SideEffectReplayContext) -> None:
...
-            side_effects.is_modified(var._base_vt)
-        )
-    if (
-        isinstance(var, variables.UserDefinedDictVariable)
-        and contents_modified
-        and var._base_vt.has_new_items()  # type: ignore[union-attr]
-    ):
+    # Dict content-modification is self-contained (items vs. original_items),
+    # so it doesn't need the mutation_type dance that list/deque still do.
+    if isinstance(var, variables.UserDefinedDictVariable) and var.has_new_items():
```
Comment (anijain2305) [inline torch/_dynamo/side_effects.py:1977]: Earlier you were handing 

```
                (variables.UserDefinedDictVariable, variables.UserDefinedSetVariable),
            )
            and item.has_new_items()
        ):
            return True

        if (
            isinstance(
                item,
                (variables.UserDefinedListVariable, variables.UserDefinedDequeVariable),
```

Seems like `UserDefinedSetVariable` is missing.

**Review by anijain2305 — COMMENTED** [review]
Looks pretty solid. Please address the AI assisted review, and `self.mutated_sources` comment.

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -1896,13 +1900,9 @@ def call_method(
             # object.__init__ is a no-op
             return variables.ConstantVariable.create(None)
 
-        if self.fn in (set, frozenset, list, tuple):
+        if self.fn in (set, frozenset, list, tuple, int, str, float, complex):
```
Comment (williamwen42) [inline torch/_dynamo/variables/builtin.py:1903]: > **AI-generated review comment - Finding A [P1]: Preserve builtin descriptor receiver checks**
>
> Forwarding only the method name to args[0] loses the descriptor's __objclass__ check. As a result, an invalid unbound call is silently reinterpreted as a valid method on the wrong receiver: int.as_integer_ratio(1.5) becomes 1.5.as_integer_ratio(), and the analogous dict path turns dict.copy([]) into [].copy().
>
> Repro:
>
> ~~~python
> import torch
> from torch._dynamo.testing import CompileCounter
>
> def int_case(x):
>     try:
>         result = int.as_integer_ratio(1.5)
>     except TypeError:
>         result = "TypeError"
>     return x + 1, result
>
> def dict_case(x):
>     try:
>         result = dict.copy([])
>     except TypeError:
>         result = "TypeError"
>     return x + 1, result
>
> for fn in (int_case, dict_case):
>     counter = CompileCounter()
>     opt_fn = torch.compile(fn, backend=counter, fullgraph=True)
>     print(fn(torch.tensor(1)))
>     print(opt_fn(torch.tensor(1)))
>     print(counter.frame_count)
> ~~~
>
> Eager catches TypeError in both cases. At PR head 068b06d66a8, the compiled results are respectively (tensor(2), (3, 2)) and (tensor(2), []), with one captured graph each. The immediate base 7304daef97a does not produce either wrong value; it conservatively raises Unsupported with no graph. The base worktree needed only its unrelated missing ConstantVariable import to load.
>
> Please restore the descriptor-based path already used by WrapperDescriptorVariable and MethodDescriptorVariable, rather than forwarding by name. In both builtin tp_getattro_impl paths, the core shape is:
>
> ~~~python
> if isinstance(attr, types.WrapperDescriptorType):
>     owner = resolve_descriptor_owner(tx, attr, self, self.fn)
>     return variables.WrapperDescriptorVariable(attr, owner=owner, source=source)
> if isinstance(attr, types.MethodDescriptorType):
>     owner = resolve_descriptor_owner(tx, attr, self, self.fn)
>     return variables.MethodDescriptorVariable(attr, owner=owner, source=source)
> ~~~
>
> Their existing call_function implementations run _check_descriptor_obj_type before invoking the defining owner. When supporting builtin subclasses through the new MI representation, bind the known builtin-base descriptor explicitly; using getattr(value, name) would incorrectly redispatch a subclass override. Please cover both invalid receivers and explicit base calls on a subclass that overrides the same method.

File: torch/_dynamo/variables/user_defined.py [hunk 4]
```diff
@@ -2664,6 +2753,8 @@ def nb_or_impl(
         reverse: bool = False,
     ) -> VariableTracker:
         # ref: https://github.com/python/cpython/blob/3.13/Objects/typeobject.c#L10354-L10355
+        if self.inherits_base_slot("__or__"):
```
Comment (williamwen42) [inline torch/_dynamo/variables/user_defined.py:2756]: > **AI-generated review comment - Finding B [P1]: Keep reflected dispatch inside SLOT1BIN**
>
> This shortcut checks only whether the forward method is inherited. For a builtin-container subclass that inherits __or__ but overrides __ror__, it delegates to the builtin VT before SLOT1BIN can apply reflected-method and subtype priority. The compiled program therefore returns the base container result instead of calling the override.
>
> Repro:
>
> ~~~python
> import torch
> from torch._dynamo.testing import CompileCounter
>
> class S(set):
>     def __ror__(self, other):
>         return "reverse"
>
> def fn(x, value):
>     return x + 1, {1} | value
>
> counter = CompileCounter()
> opt_fn = torch.compile(fn, backend=counter, fullgraph=True)
> print(fn(torch.tensor(1), S({2})))
> print(opt_fn(torch.tensor(1), S({2})))
> print(counter.frame_count)
> ~~~
>
> Immediate base 7304daef97a returns (tensor(2), "reverse"); PR head 068b06d66a8 returns (tensor(2), {1, 2}) with one captured graph. The same failure affects reflected set operations and reflected addition on list, tuple, and deque subclasses. It also occurs for BaseSet() | SubSet() when only SubSet.__ror__ is overridden, so changing this check to __ror__ only when reverse=True is insufficient.
>
> A targeted fix is to make inherited C descriptors terminate through the MI base inside _vectorcall_maybe:
>
> ~~~python
> if self.inherits_base_slot(name):
>     return self.call_base_method(tx, name, args, {})
> ~~~
>
> Then remove the eager inherits_base_slot(forward_dunder) returns from the two-argument nb_*_impl methods and let SLOT1BIN perform Python's forward/reflected/subtype ordering. The inherited-descriptor branch above prevents recursion. Keep the direct base path for three-argument pow, where reflected dispatch does not apply. Please add reverse-only tests with both a builtin LHS and a two-level subclass hierarchy.

File: torch/_dynamo/side_effects.py [hunk 5]
```diff
@@ -1972,28 +1984,26 @@ def _codegen_attribute_mutation(ctx: SideEffectReplayContext) -> None:
...
+    elif isinstance(var, variables.UserDefinedSetVariable) and var.has_new_items():
+        _codegen_user_defined_set_mutation(ctx)
+    elif isinstance(var, variables.DefaultDictVariable) and var.has_new_items():
+        _codegen_const_dict_or_set_mutation(ctx)
+    else:
+        mt = var.mutation_type
+        if isinstance(mt, AttributeMutationNew):
+            contents_modified = True
+        elif isinstance(mt, AttributeMutationExisting):
+            contents_modified = side_effects.is_modified(var)
```
Comment (williamwen42) [inline torch/_dynamo/side_effects.py:2000]: > **AI-generated review comment - Finding C [P1]: Do not replay deque contents for attribute-only mutations**
>
> ValueAndAttributeMutationExisting has independent content and attribute mutation state, but side_effects.is_modified(var) combines both axes. Consequently, setting only an instance attribute makes this branch clear and extend an otherwise unchanged deque. Those operations increment the deque's internal mutation state and invalidate iterators that eager execution leaves valid.
>
> Repro:
>
> ~~~python
> import collections
> import torch
> from torch._dynamo.testing import CompileCounter
>
> class D(collections.deque):
>     pass
>
> def fn(x, value):
>     value.marker = "set"
>     return x + 1
>
> value = D([1, 2])
> iterator = iter(value)
> counter = CompileCounter()
> opt_fn = torch.compile(fn, backend=counter, fullgraph=True)
> opt_fn(torch.tensor(1), value)
> print(list(value))
> print(next(iterator))
> print(counter.frame_count)
> ~~~
>
> On immediate base 7304daef97a, this prints [1, 2], 1, and 1. At PR head 068b06d66a8, next(iterator) raises RuntimeError: deque mutated during iteration, despite the contents still being [1, 2] and one graph being captured.
>
> The existing-container branch should consult only the value-mutation flag while retaining full materialization for newly created containers:
>
> ~~~diff
>  if isinstance(mt, AttributeMutationNew):
>      contents_modified = True
> -elif isinstance(mt, AttributeMutationExisting):
> -    contents_modified = side_effects.is_modified(var)
> +elif isinstance(mt, ValueMutationExisting):
> +    contents_modified = mt.is_modified
>  else:
>      contents_modified = False
> ~~~
>
> Please add an attribute-only deque-subclass test that creates an iterator before the compiled call and checks that next(iterator) still returns the first element.

**Comment by williamwen42** [comment]
AI found 3 corner cases that look legitimate; otherwise, the changes look good to me.
